### PR TITLE
Rewrite file-sniffer analysis

### PR DESCRIPTION
**Pull request by angelikatyborska**
Resolves https://github.com/exercism/elixir-analyzer/issues/407

Originally this PR was called:
> Rewrite file-sniffer analysis (assert_call should also find calls in function heads)

After a deep dive, I realized that `assert_call` is already capable of finding calls in function heads.

I have initially made a wrong assumption. I thought that the `Kernel` module alone is implicitly imported into each new module. That is only partially correct. The `Kernel.SpecialForms` module is also implicitly imported, with some important differences:
- You cannot modify or opt out of importing `Kernel.SpecialForms`, whereas you can modify `Kernel` imports by making them explicit and passing `:only`/`:except` options
- In the AST, usages of Kernel functions can be distinguished from local calls, but usages of `Kernel.SpecialForms` macros cannot. They look exactly like usages of local function/macros.

I was also confused which module provides which macro, I thought `<<>>` and `::` comes from `Kernel`, but it coems from `Kernel.SpecialForms`

**TL;DR**: we can use `assert_call called_fn name: :<<>>` to check for calls of special forms with no problems and no modifications necessary. It is not allowed to overwrite those special forms anyway, so you won't have a conflicting local function.

Related PR: https://github.com/exercism/website-copy/pull/2306

### CODE REVIEW

**Comment by angelikatyborska** [comment]
@jiegillet could you review this one?

File: test/elixir_analyzer/test_suite/file_sniffer_test.exs [hunk 1]
```diff
@@ -147,40 +153,90 @@ defmodule ElixirAnalyzer.ExerciseTest.FileSnifferTest do
...
+
+        def type_from_binary(<<127, 69, 76, 70, _::binary>>), do: type_from_extension("exe")
+        def type_from_binary(file), do: type_from_extension(file)
       end
     ]
   end
 
-  test_exercise_analysis "doesn't use pattern matching",
-    comments: [Constants.file_sniffer_use_pattern_matching()] do
+  test_exercise_analysis "doesn't use <<>> nor ::",
```
Comment (jiegillet) [inline test/elixir_analyzer/test_suite/file_sniffer_test.exs:210]: The comment should be triggered if the solution only uses one of the forms right? Should we add tests to check that?

**Review by jiegillet — APPROVED** [review]
Wonderful, it's nice to see that `assert_call` was already powerful enough to handle that, thank you for adding the new tests to highlight the behavior.

Sorry that I took so long to review, I still thought that `assert_call` needed to change, and I was slowly gathering motivation to deep dive back into it, but it seems I never had enough haha. I'm glad you reminded me :)

The code looks great, I just left a suggestion to add a test or two, but it's just a formality.

**Comment by jiegillet** [comment]
GitHub doesn't seem to let me re-run the external tests, I can't find the button, I don't know why.

File: test/elixir_analyzer/test_suite/file_sniffer_test.exs (hunk 1 above)
Comment (angelikatyborska) [reply test/elixir_analyzer/test_suite/file_sniffer_test.exs:210]: I added a test with just `<<>>` but I don't think I can add a test with just `::`. It's used for defining binary sizes, so you can't really use it without using the binary special form...

**Comment by angelikatyborska** [comment]
>  I can't find the button, I don't know why.

You can only rerun the tests if the latest test run isn't too old... I think maybe 2 or 4 weeks.

**Comment by jiegillet** [comment]
Ah that makes sense :)
